### PR TITLE
Adds the Web on Android blogpost

### LINKED ISSUE #1313: content: When do I use PWA/TWA/WebView/CCT
**Author**: @andreban
**Reviewer**: @jpmedley

### PR DESCRIPTION
**Pull request by andreban**
- Adds the Web on Android blogpost

@jpmedley This is sitting inside the `progressive-web-apps` folder, but I'm not sure if that's the best place for it. I have added a `DO NOT MERGE` label just to make sure we confirm the appropriate place before merging. Let me know what you think.

Closes #1313

### CODE REVIEW

**Comment by jpmedley** [comment]
@andreban I will start the edit. I think the question about where this lives is for @petele.

File: src/site/content/en/progressive-web-apps/web-on-android/index.md [hunk 1]
```diff
@@ -0,0 +1,142 @@
+---
+layout: post
+title: Web on Android
+authors:
+  - andreban
+date: 2020-07-30
+updated: 2020-07-30
+description: |
+  Learn about how different components can be used to render web content inside Android apps.
```
Comment (jpmedley) [inline src/site/content/en/progressive-web-apps/web-on-android/index.md:9]:
```suggestion
  Learn how different components can be used to render web content inside Android apps.
```

File: src/site/content/en/progressive-web-apps/web-on-android/index.md [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+## Web on Android
+
+The Android platform has been around for more than ten years, and since its early days it has had
+great support for the Web. It shipped with WebView, a component that allows developers to use the
+web inside their own Android Apps. More than that, Android allows developers to bring their own
+browser engine into the platform, fostering competition and innovation.
+
+Developers can include the web in their Android applications in many ways. WebView is frequently
+used to render ads, as a layout component used along with native UI elements, or packaging HTML 5
```
Comment (jpmedley) [inline src/site/content/en/progressive-web-apps/web-on-android/index.md:22]:
```suggestion
used to render ads, as a layout component used along with native UI elements, or for packaging HTML 5
```

File: src/site/content/en/progressive-web-apps/web-on-android/index.md [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+web platform features, developers would still need to write code and implement their own UI for
+functionality like permissions or push notifications, making it hard to achieve consistency for
+users. 
+
+Another option available to developers is embedding a browser engine in their application. Besides
+leading to increased application size, this approach is both complex and time-consuming.
+
+### Custom Tabs as a solution for in-app browsers
+
+Custom Tabs were [introduced in Chrome 45][2] and allows developers to use a tab from the user's
```
Comment (jpmedley) [inline src/site/content/en/progressive-web-apps/web-on-android/index.md:62]:
```suggestion
Custom Tabs was [introduced in Chrome 45][2] and allows developers to use a tab from the user's
```

File: src/site/content/en/progressive-web-apps/web-on-android/index.md [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+default browser as part of their application. Custom Tabs were originally launched by Chrome, and
```
Comment (jpmedley) [inline src/site/content/en/progressive-web-apps/web-on-android/index.md:63]:
```suggestion
default browser as part of their application. Custom Tabs was originally launched by Chrome, and
```

File: src/site/content/en/progressive-web-apps/web-on-android/index.md [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+were therefore known as "Chrome Custom Tabs|. Today it's an [Android API][3] and most popular
```
Comment (jpmedley) [inline src/site/content/en/progressive-web-apps/web-on-android/index.md:64]:
```suggestion
was therefore known as "Chrome Custom Tabs". Today it's an [Android API][3] and most popular
```

File: src/site/content/en/progressive-web-apps/web-on-android/index.md [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+browsers support Custom Tabs, including Chrome, Firefox, Edge and Samsung Internet, so it's more
+appropriate to just call them "Custom Tabs".
+
+Custom Tabs help developers seamlessly integrate web content into their app experience. They also
```
Comment (jpmedley) [inline src/site/content/en/progressive-web-apps/web-on-android/index.md:68]:
```suggestion
Custom Tabs helps developers seamlessly integrate web content into their app experience. They also
```

File: src/site/content/en/progressive-web-apps/web-on-android/index.md [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+allow developers to customise the activity in which web content is shown by allowing them to
+customize the toolbar color, action buttons, transition animation, and more.
+
+They also offer features that were previously unavailable when using WebView or embedding a browser
+engine. Since the in-app browser is powered by the user's browser, a Custom Tabs shares storage
+with the browser, so users don't need to re-login to their favourite websites every time one of
+their installed apps starts an In-App browsing session.
+
+Unlike WebViews, Custom Tabs support all web platform features and APIs that are supported by the
```
Comment (jpmedley) [inline src/site/content/en/progressive-web-apps/web-on-android/index.md:77]:
```suggestion
Unlike WebViews, Custom Tabs supports all web platform features and APIs that are supported by the
```

File: src/site/content/en/progressive-web-apps/web-on-android/index.md [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+performance characteristics, Trusted Web Activities introduce a [quality criteria][8] for PWAs
+being opened inside it.
+
+### Limitations of the current solutions
+
+Developer feedback showed a need for the platform compatibility of Custom Tabs combined with the
+flexibility of WebView so they could, for instance, access the DOM or inject JavaScript, into their
+in-app browsers.
+
+Custom Tabs are effectively a tab rendered by the user's browser, with a custom UI or with no UI
```
Comment (jpmedley) [inline src/site/content/en/progressive-web-apps/web-on-android/index.md:112]:
```suggestion
Custom Tabs is effectively a tab rendered by the user's browser, with a custom UI or with no UI
```

File: src/site/content/en/progressive-web-apps/web-on-android/index.md [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+Notifications, Web Bluetooth and others. It is not recommended when opening content that has been
+designed for the modern web platform, as it may not be displayed in the way the developer intended.
+WebView is not recommended for creating in-app browsers. On the other hand displaying first-party
+web content is an area where WebViews really shine. 
+
+Trusted Web Activity should be used when the developers want to render their own Progressive Web
+App in fullscreen, inside their Android application. It can be used as the only activity in the app
+or used along with other native activities.
+
+Custom Tabs are the recommended way for opening third-party content that is designed for the web
```
Comment (jpmedley) [inline src/site/content/en/progressive-web-apps/web-on-android/index.md:132]:
```suggestion
Custom Tabs is the recommended way for opening third-party content that is designed for the web
```

File: src/site/content/en/progressive-web-apps/web-on-android/index.md [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+used to render ads, as a layout component used along with native UI elements, or packaging HTML 5
+games. Custom Tabs allows developers to build in-app browsers and provide a seamless navigation
+experience to third-party web content, and Trusted Web Activity allows developers to use their
+Progressive Web Apps (PWAs) in Android apps, which can be downloaded from the Play Store.
+
+### Android WebView
+
+WebView gives developers access to modern HTML, CSS, and JavaScript inside their Android apps, and
+allows content to be shipped inside the APK or hosted on the internet. It's one of Android's most
+flexible and powerful components that can be used for most of the use-cases where web content is
```
Comment (jpmedley) [inline src/site/content/en/progressive-web-apps/web-on-android/index.md:31]:
```suggestion
flexible and powerful components, which can be used for most of the use-cases where web content is
```

File: src/site/content/en/progressive-web-apps/web-on-android/index.md [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+included in an Android app. From powering ad services, like AdMob to building and shipping complete
```
Comment (jpmedley) [inline src/site/content/en/progressive-web-apps/web-on-android/index.md:32]:
```suggestion
included in an Android app. From powering ad services like AdMob to building and shipping complete
```

File: src/site/content/en/progressive-web-apps/web-on-android/index.md [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+HTML5 games that use modern APIs, such as WebGL.
```
Comment (jpmedley) [inline src/site/content/en/progressive-web-apps/web-on-android/index.md:33]:
```suggestion
HTML5 games that use modern APIs such as WebGL.
```

File: src/site/content/en/progressive-web-apps/web-on-android/index.md [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+
+But, when used to create an in-app-browser or including a PWA in an Android application, WebView
+lacks the security, features, and capabilities of the web platform
```
Comment (jpmedley) [inline src/site/content/en/progressive-web-apps/web-on-android/index.md:36]:
```suggestion
lacks the security, features, and capabilities of the web platform.
```

File: src/site/content/en/progressive-web-apps/web-on-android/index.md [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+
+### The in-app browser challenge
+
+Over time, the frequency of developers building browser experiences for third-party content into
+their Android application increased, with the goal of creating a more seamless
+experience for their users when navigating to third-party websites. Those experiences became known
```
Comment (jpmedley) [inline src/site/content/en/progressive-web-apps/web-on-android/index.md:42]:
```suggestion
Over time, more and more developers have built brower experiences 
incorporating third-party content into their Android application, with the goal
of creating a more seamless experience for their users when navigating to
third-party websites. Those experiences became known
```

File: src/site/content/en/progressive-web-apps/web-on-android/index.md [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+
+Another option available to developers is embedding a browser engine in their application. Besides
+leading to increased application size, this approach is both complex and time-consuming.
+
+### Custom Tabs as a solution for in-app browsers
+
+Custom Tabs were [introduced in Chrome 45][2] and allows developers to use a tab from the user's
+default browser as part of their application. Custom Tabs were originally launched by Chrome, and
+were therefore known as "Chrome Custom Tabs|. Today it's an [Android API][3] and most popular
+browsers support Custom Tabs, including Chrome, Firefox, Edge and Samsung Internet, so it's more
```
Comment (jpmedley) [inline src/site/content/en/progressive-web-apps/web-on-android/index.md:65]:
```suggestion
browsers support Custom Tabs, including Chrome, Firefox, Edge, and Samsung Internet, so it's more
```

File: src/site/content/en/progressive-web-apps/web-on-android/index.md [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+appropriate to just call them "Custom Tabs".
```
Comment (jpmedley) [inline src/site/content/en/progressive-web-apps/web-on-android/index.md:66]:
```suggestion
appropriate to just call it "Custom Tabs".
```

File: src/site/content/en/progressive-web-apps/web-on-android/index.md [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+
+Custom Tabs help developers seamlessly integrate web content into their app experience. They also
+allow developers to customise the activity in which web content is shown by allowing them to
+customize the toolbar color, action buttons, transition animation, and more.
+
+They also offer features that were previously unavailable when using WebView or embedding a browser
+engine. Since the in-app browser is powered by the user's browser, a Custom Tabs shares storage
```
Comment (jpmedley) [inline src/site/content/en/progressive-web-apps/web-on-android/index.md:73]:
```suggestion
engine. Since the in-app browser is powered by the user's browser, Custom Tabs shares storage
```

File: src/site/content/en/progressive-web-apps/web-on-android/index.md [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+browser powering it.
+
+### Open Progressive Web Apps using Trusted Web Activity
+
+[Progressive Web Apps][4] bring many behaviours and capabilities that were once only available to
+native apps to the web. With the introduction of app-like behaviour, the desire from developers to
+re-use those experiences on Android increased, and developers started asking for ways to integrate
+PWAs into their apps.
+
+Custom Tabs have support for all modern web capabilities and APIs but, since it was primarily
```
Comment (jpmedley) [inline src/site/content/en/progressive-web-apps/web-on-android/index.md:87]:
```suggestion
Custom Tabs has support for all modern web capabilities and APIs but, since it was primarily
```

File: src/site/content/en/progressive-web-apps/web-on-android/index.md [hunk 19]
```diff
@@ -0,0 +1,142 @@
...
+designed to open third-party content, it has a toolbar on the top that tells the users which URL
+they are visiting, as well as the lock icon indicating if the site is secure. When opening an app's
```
Comment (jpmedley) [inline src/site/content/en/progressive-web-apps/web-on-android/index.md:89]:
```suggestion
they are visiting, as well as the lock icon indicating whether the site is secure. When opening an app's
```

File: src/site/content/en/progressive-web-apps/web-on-android/index.md [hunk 20]
```diff
@@ -0,0 +1,142 @@
...
+own experience, the toolbar prevents the application from feeling like it is integrated with the
+operating system.
+
+[Trusted Web Activities][5] were introduced in Chrome 72 and allow developers to
```
Comment (jpmedley) [inline src/site/content/en/progressive-web-apps/web-on-android/index.md:93]:
```suggestion
[Trusted Web Activities][5] was introduced in Chrome 72 and allows developers to
```

File: src/site/content/en/progressive-web-apps/web-on-android/index.md [hunk 21]
```diff
@@ -0,0 +1,142 @@
...
+[use their PWA inside an Android][6] app. Their protocol is similar to the Custom Tabs protocol,
```
Comment (jpmedley) [inline src/site/content/en/progressive-web-apps/web-on-android/index.md:94]:
```suggestion
[use their PWA inside an Android][6] app. Its protocol is similar to the Custom Tabs protocol,
```

File: src/site/content/en/progressive-web-apps/web-on-android/index.md [hunk 22]
```diff
@@ -0,0 +1,142 @@
...
+but introduces APIs that allow developers to verify (through [Digital Asset Links][7]) that they
+control both the Android app and the URL being opened and remove the URL bar when both are true.
+
+They also introduced APIs for creating splash screens when opening the PWA or delegating web
+notifications to be handled by native code. More features like support for Play Billing are coming
+soon.
+
+Since URLs opened in Trusted Web Activities are expected to be PWAs and have a set of behaviors and
+performance characteristics, Trusted Web Activities introduce a [quality criteria][8] for PWAs
```
Comment (jpmedley) [inline src/site/content/en/progressive-web-apps/web-on-android/index.md:103]:
```suggestion
performance characteristics, Trusted Web Activities introduces [quality criteria][8] for PWAs
```

File: src/site/content/en/progressive-web-apps/web-on-android/index.md [hunk 23]
```diff
@@ -0,0 +1,142 @@
...
+Custom Tabs are effectively a tab rendered by the user's browser, with a custom UI or with no UI
+at all. This means that the browser needs to honour the user's expectations around privacy and
+security towards the browser, making some of those features impossible.
+
+The Web on Android team at Google is looking into alternatives and experimenting with solutions
+to solve those use-cases. Stay tuned for details!
+
+### Summary
+
+WebView is useful when an application needs HTML, CSS and JavaScript inside their Android app, but
```
Comment (jpmedley) [inline src/site/content/en/progressive-web-apps/web-on-android/index.md:121]:
```suggestion
WebView is useful when an application needs HTML, CSS, and JavaScript inside their Android app, but
```

File: src/site/content/en/progressive-web-apps/web-on-android/index.md [hunk 24]
```diff
@@ -0,0 +1,142 @@
...
+doesn't use more advanced features and capabilities available on the modern web, such as Web Push
```
Comment (jpmedley) [inline src/site/content/en/progressive-web-apps/web-on-android/index.md:122]:
```suggestion
doesn't use more advanced features and capabilities available on the modern web such as Push
```

File: src/site/content/en/progressive-web-apps/web-on-android/index.md [hunk 25]
```diff
@@ -0,0 +1,142 @@
...
+Notifications, Web Bluetooth and others. It is not recommended when opening content that has been
+designed for the modern web platform, as it may not be displayed in the way the developer intended.
+WebView is not recommended for creating in-app browsers. On the other hand displaying first-party
+web content is an area where WebViews really shine. 
+
+Trusted Web Activity should be used when the developers want to render their own Progressive Web
+App in fullscreen, inside their Android application. It can be used as the only activity in the app
```
Comment (jpmedley) [inline src/site/content/en/progressive-web-apps/web-on-android/index.md:129]:
```suggestion
App in fullscreen inside their Android application. It can be used as the only activity in the app
```

File: src/site/content/en/progressive-web-apps/web-on-android/index.md [hunk 26]
```diff
@@ -0,0 +1,142 @@
...
+  - andreban
+date: 2020-07-30
+updated: 2020-07-30
+description: |
+  Learn about how different components can be used to render web content inside Android apps.
+tags:
+  - android
+---
+
+## Web on Android
```
Comment (kaycebasques) [inline src/site/content/en/progressive-web-apps/web-on-android/index.md:14]:
```suggestion
```

Just a drive-by redundancy that I caught while browsing this PR

image

**Comment by kaycebasques** [comment]
@andreban is this the content related to #2436 or #1313?

**Comment by andreban** [comment]
@kaycebasques this content is related to #1313. Updated the description.

**Comment by jpmedley** [comment]
@petele Where should this go? I would prefer to add the updated json file at the same time.

**Comment by petele** [comment]
I think it's probably better for this to fit under `/blog/` for now, since it's not PWA specific.

**Comment by jpmedley** [comment]
@andreban, the ball is in your court.

**Comment by andreban** [comment]
@jpmedley moved to blog. Also updated the tag to be `blog`, as it seems required for the blog.

File: src/site/content/en/blog/web-on-android/index.md [hunk 27]
```diff
@@ -8,7 +8,7 @@ updated: 2020-07-30
 description: |
   Learn how different components can be used to render web content inside Android apps.
 tags:
-  - android
+  - blog # blog is a required tag for the article to show up in the blog.
```
Comment (jpmedley) [inline src/site/content/en/blog/web-on-android/index.md:11]:
```suggestion
  - android
  - blog # blog is a required tag for the article to show up in the blog.
```

Comment (jpmedley) [reply src/site/content/en/blog/web-on-android/index.md:11]: More than one tag is allowed.

Comment (andreban) [reply src/site/content/en/blog/web-on-android/index.md:12]: Awesome, thank you. Commited!

**Comment by jpmedley** [comment]
@andreban You're the one who tagged this 'do not merge'. Are we waiting on anything?

**Comment by andreban** [comment]
@jpmedley sorry for that. Removed the `DO NOT MERGE` tag.

**Comment by jpmedley** [comment]
@robdodson Changed the CLA tag manually on this because all the contributors are Google employees. Not sure if that's the right way to do it.

**Comment by robdodson** [comment]
> @robdodson Changed the CLA tag manually on this because all the contributors are Google employees. Not sure if that's the right way to do it.

yep that's the way to do it.
